### PR TITLE
feat!: 6 new commands to toggle text between `PascalCase`, `camelCase`, `snake_case`, `Title Case`,  `kebab-case` and `Sentence case`

### PR DESCRIPTION
**Pull request by nik-rev**
This PR introduces a new mode under `` ` `` called "case" mode which has several utility key mappings for working with switching case of text

`switch_case`, `switch_to_lowercase` and `switch_to_uppercase` have been moved to this new mode. 6 new case conversion commands have been added and are accessible under `` ` `` as well:

| Key              | Description                                      | Command                    |
| -----            | -----------                                      | -------                    |
| `` `l ``              | Switch text to lowercase                     | `switch_to_lowercase`      |
| `` `u ``              | Switch text to UPPERCASE                     | `switch_to_uppercase`      |
| `` `p ``              | Switch text to PascalCase                       | `switch_to_pascal_case`    |
| `` `c ``              | Switch text to camelCase                         | `switch_to_camel_case`     |
| `` `t ``              | Switch text to Title Case                        | `switch_to_title_case`     |
| `` `S ``              | Switch text to Sentence case                        | `switch_to_sentence_case`     |
| `` `s ``              | Switch text to snake_case                        | `switch_to_snake_case`     |
| `` `k ``              | Switch text to kebab-case                        | `switch_to_kebab_case`     |
| `` `a ``              | Switch text to aLTERNATE cASE                     | `switch_to_alternate_case`      |

Closes https://github.com/helix-editor/helix/issues/5197

---

Unlike https://github.com/helix-editor/helix/pull/4615 this PR doesn't pull in any extra dependencies. It builds on top of the existing `case_conversion.rs` module and adds many tests to make sure that functionality is correct. The tests even fixed some edges with the existing `camelCase` implementation

## Breaking changes

Remaps two keys:

|old|new|
|-|-|
|`` Alt-` ``|`` `u ``|
| `` ` ``|`` `l ``|

Renames a command, for consistency:

|old|new|
|-|-|
|`switch_case`|`switch_to_alternate_case`|

### CODE REVIEW

**Review by TornaxO7 — COMMENTED** [review]
Interesting feature idea, however, I'm a bit unsure if this might be more suitable as a plugin than as a core feature.

**Comment by nikitarevenco** [comment]
@TornaxO7 it looks like you may have marked your comment as off-topic. Was this intended?

**Comment by NikitaRevenco** [comment]
> Interesting feature idea, however, I'm a bit unsure if this might be more suitable as a plugin than as a core feature.

Imo, it shouldn't be a plugin. This is a useful feature that is handy to have in an editor when you need it, but it's not something you know you need until you need it.

**Comment by NikitaRevenco** [comment]
> Interesting feature idea, however, I'm a bit unsure if this might be more suitable as a plugin than as a core feature.

Btw, I only replied now because your reply was marked as "off topic" before

How do comments get marked as off-topic? Is it automatic ,or someone marked yours? not sure how it works

**Comment by TornaxO7** [comment]
> How do comments get marked as off-topic?

You can mark your own comments as off-topic but I also think that the repo owners are able to mark comments as offtopic as well.

> Is it automatic ,or someone marked yours?

I did it myself because I thought that my comment there might be not that "helpful" so I just marked it as "offtopic", just to be safe :)

File: helix-term/Cargo.toml [hunk 1]
```diff
@@ -35,6 +35,7 @@ helix-loader = { path = "../helix-loader" }
 anyhow = "1"
 once_cell = "1.20"
 
+heck = "0.5"
```
Comment (filipdutescu) [inline helix-term/Cargo.toml:38]: Looks very nice, I think this is the only thing that maintainers might want to consider, since it's adding a new dependency. Otherwise I am very impressed with how small the changes are! Congrats!

File: book/src/keymap.md [hunk 2]
```diff
@@ -232,6 +232,24 @@ Jumps to various locations.
 | `k`   | Move up textual (instead of visual) line         | `move_line_up`             |
 | `w`   | Show labels at each word and select the word that belongs to the entered labels | `goto_word` |
 
+#### Case mode
+
+Accessed by typing ` ` ` in [normal mode](#normal-mode).
+
+Various functions for changing case of text in different ways.
```
Comment (filipdutescu) [inline book/src/keymap.md:239]: Nit:

```suggestion
Various commands for changing the case of text in different ways.
```

**Review by filipdutescu — APPROVED** [review]

**Comment by gabydd** [comment]
We have case conversion in helix core from the snippets system which should be used instead of another dependency https://github.com/helix-editor/helix/blob/master/helix-core%2Fsrc%2Fcase_conversion.rs

File: helix-term/Cargo.toml (hunk 1 above)
Comment (NikitaRevenco) [reply helix-term/Cargo.toml:38]: Thanks! I'm gonna use Helix's built-in case conversions so the dependency is removed

**Comment by darshanCommits** [comment]
Hopefully this gets merged, now that it's not even pulling dep.

File: helix-core/src/case_conversion.rs [hunk 3]
```diff
@@ -2,27 +2,45 @@ use crate::Tendril;
...
 // todo: should this be grapheme aware?
 
-pub fn to_pascal_case(text: impl Iterator<Item = char>) -> Tendril {
+fn to_case<I>(text: I, to_case_with: fn(I, &mut Tendril)) -> Tendril
+where
+    I: Iterator<Item = char>,
+{
     let mut res = Tendril::new();
-    to_pascal_case_with(text, &mut res);
+    to_case_with(text, &mut res);
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:10]: Nit: i don't like the `_with` suffix too much, I find it doesn't convey much meaning, at least in my opinion. I would much rather have it named something like 


```suggestion
    case_conversion(text, &mut res);
```

Or

```suggestion
    convert_to_case(text, &mut res);
```

Or something like that. Then the other methods could be given the respective `_with` methods as callbacks:

```rust
to_case(text, |args...| { /* Implementation... */ })
```

Comment (NikitaRevenco) [reply helix-core/src/case_conversion.rs]: All the functions which mutate in-place I've prefixed with `into`, all the ones which don't are prefixed with `to` now. please let me know what you think now!

**Comment by NikitaRevenco** [comment]
Okay, finally refactored everything. It doesn't need dependency anymore. I've added a lot of tests. The tests even caught a bug with the original implementation for camelCase, which has been fixed

File: helix-core/src/case_conversion.rs [hunk 4]
```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
...
+    separator: char,
+) {
+    let mut prev: Option<char> = None;
+
+    for c in text.skip_while(|ch| ch.is_whitespace()) {
+        if c.is_alphanumeric() {
+            if prev.is_some_and(|p| p.is_lowercase()) && c.is_uppercase()
+                || !prev.is_some_and(|p| p.is_alphanumeric()) && !buf.is_empty()
+            {
+                buf.push(separator);
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:60]: Nit: invert `if` and use `continue` so you need to focus on less nesting when reading

File: helix-core/src/case_conversion.rs [hunk 5]
```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
...
+    capitalize_first: bool,
+    separator: Option<char>,
+) {
+    let mut capitalize_next = capitalize_first;
+    let mut prev: Option<char> = None;
+
+    for c in text.skip_while(|ch| ch.is_whitespace()) {
+        if c.is_alphanumeric() {
+            if prev.is_some_and(|p| p.is_lowercase()) && c.is_uppercase() {
+                capitalize_next = true;
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:28]: Would this not be better named `capitalize_current`? Seems like you're capitalising the current char

File: helix-core/src/case_conversion.rs [hunk 6]
```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
...
+            }
         } else {
-            buf.push(c)
+            capitalize_next = true;
+            if let Some(separator) = separator {
+                if prev.is_some_and(|p| p != separator) {
+                    buf.push(separator);
+                }
+            }
         }
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:43]: I would merge the conditions into one `if`, unless the compiler complains

File: helix-core/src/case_conversion.rs [hunk 7]
```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
...
+
+pub fn into_snake_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    separator_case_conversion(text, buf, '_');
+}
+
+pub fn into_title_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
+    complex_case_conversion(text, buf, true, Some(' '));
+}
+
+pub fn into_camel_case(text: impl Iterator<Item = char>, buf: &mut Tendril) {
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:101]: Since we have Title Case we should probably include Sentence case as well

File: helix-core/src/case_conversion.rs [hunk 8]
```diff
@@ -2,68 +2,302 @@ use crate::Tendril;
...
+
+        title_test("hello  world", "Hello World");
+
+        title_test("   hello world", "Hello World");
+        title_test("hello\tworld", "Hello World");
+        // title_test("HELLO  WORLD", "Hello World");
+        title_test("HELLO-world", "Hello World");
+        // title_test("hello  WORLD ", "Hello World");
+        // title_test("helloWorld", "Hello World");
+    }
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:268]: Any particular reason those are commented? They should probably be enabled

**Comment by filipdutescu** [comment]
Excellent work! Left a few more comments with minor possible improvements

File: helix-core/src/case_conversion.rs [hunk 9]
```diff
@@ -2,68 +2,328 @@ use crate::Tendril;
...
+    to_case(text, into_snake_case)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_camel_case_conversion() {
+        let tests = [
```
Comment (filipdutescu) [inline helix-core/src/case_conversion.rs:156]: Maybe, if there are others in Helix already, we could use parameterised tests? I know it requires some extra dependency, so if there are no other such tests, just resolve this comment.

File: helix-core/src/case_conversion.rs (hunk 7 above)
Comment (NikitaRevenco) [reply helix-core/src/case_conversion.rs:111]: I don't think it's very useful. Maybe I'm wrong, but also `s` is reserved for snake_case and what key would be reserved for sentence case? S wouldn't make sense since usually a lowercase letter and capital letter will do opposite actions to each other like `p` and `P`. So we would have `s` for snake_case and `S` for sentence case, when they're not really related 

I just think consistency is important. but if you have some good ideas let me know!

File: helix-core/src/case_conversion.rs (hunk 4 above)
Comment (NikitaRevenco) [reply helix-core/src/case_conversion.rs]: Ok

File: helix-core/src/case_conversion.rs (hunk 6 above)
Comment (NikitaRevenco) [reply helix-core/src/case_conversion.rs:51]: unfortunately:
```
`let` expressions in this position are unstable
see issue #53667 <https://github.com/rust-lang/rust/issues/53667> for more information
```

File: helix-core/src/case_conversion.rs (hunk 5 above)
Comment (NikitaRevenco) [reply helix-core/src/case_conversion.rs]: yes, good spot

File: helix-core/src/case_conversion.rs (hunk 9 above)
Comment (NikitaRevenco) [reply helix-core/src/case_conversion.rs:174]: doesn't seem to be the case

**Review by filipdutescu — APPROVED** [review]

**Review by filipdutescu — APPROVED** [review]

File: helix-core/src/case_conversion.rs [hunk 10]
```diff
@@ -7,9 +7,7 @@ pub fn simple_case_conversion(
     buf: &mut Tendril,
     transform_char: impl Fn(&char) -> char,
 ) {
-    for c in text {
-        buf.push(transform_char(&c))
-    }
+    *buf = text.map(|ch| transform_char(&ch)).collect();
```
Comment (darshanCommits) [inline helix-core/src/case_conversion.rs]: ```*buf = text.map(transform_char).collect();```
should work just fine.

Comment (nik-rev) [reply helix-core/src/case_conversion.rs]: Yeah I tried. It's not trivial to make this work without changing a bunch of other code

**Comment by nik-rev** [comment]
I added back the `~` mapping, as I realized it's too much of a breaking change to remove it (Vim had it for decades, and Helix...)

**Comment by yerlaser** [comment]
Thanks for this very useful feature.

I've noticed one small issue - if a word is at the end of the line and the cursor is at the new line character, then the word gets selected with `b` then the new line character also gets selected.
And then the commands to change the case (e.g. Title Case) remove the new line character.

I've compared it with the built-in Uppercase command and that doesn't happen.

**Comment by David-Else** [comment]
This looks cool, but it does not seem to do complete title case conversion: https://en.wikipedia.org/wiki/Title_case

While there are different definitions of exactly what title case is for different writing styles, they all basically keep minor words lower case. I am not sure if there is an open source crate to borrow this code for English from, and also it would be hard to make it multilingual.

**Comment by nik-rev** [comment]
Yeah, that's unfortunate..

I made this PR a fairly long time ago - I initially used the `heck` dependency to do case conversion and it was fairly low maintenance. But Helix isn't too keen on adding new dependencies, which is understandable

Several months later, looking at this again - I personally don't see the benefit of re-implementing this functionality manually and maintaining essentially a copy of a crate which has had many years to iron out the edge cases. Essentially re-solving problems that have already been solved

I'm unfortunately going to close this PR for the above reason

If you're is still interested in this feature, check out the commit https://github.com/helix-editor/helix/commit/f88364329a8024bd9218a2759ea689ccc7d3c4af which is before I removed the dependency on `heck`. The functionality should consider all edge-cases

if someone else wants to make this PR to Helix and it get accepted (as helix are not interested in new deps) - they'll need to replicate Heck's test suite and make sure it aligns with the re-implementation in Helix, so the case conversion commands are correct.

**Comment by rockboynton** [comment]
Damn, I loved this PR and use it all the time!

**Comment by Smithx10** [comment]
Yikes.  This was very useful.
